Add search service

heger2 opened the pull request.

heger2 committed:
Add search service

heger2 committed:
Add run instructions

brandon-ja commented on server.py:
```diff
@@ -0,0 +1,32 @@
+import githubsearch_pb2 as schema
+import githubsearch_pb2_grpc
+
+import utils
```
this name is pretty generic and seems like it's an external library rather than code specific to this service. Perhaps consider renaming this. Seems like this name could cause conflict with other code at some point.

heger2 replied:
I renamed this to githubsearch

brandon-ja commented on server.py:
```diff
@@ -0,0 +1,32 @@
+
+import utils
+
+SERVER_PORT = os.environ.get("GITHUB_SEARCH_SERVICE_PORT", "50051")
```
why use an env var here rather than a package constant? I don't see anything in the readme about this.

heger2 replied:
This is now able to be set via a cli arg, env var, or yaml file

brandon-ja commented on server.py:
```diff
@@ -0,0 +1,32 @@
+        return response
+
+def serve():
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
```
love the threadpool here

brandon-ja commented on utils.py:
```diff
@@ -0,0 +1,56 @@
+def _check_search_validity(search_term):
+    # The github api does not support:
+    # - longer than 256 char (not including operators or qualifiers i.e. just the search phrase)
+    # - more than five AND, OR, or NOT operators
```
great comment. also, include a link to the github docs here

heger2 replied:
link added :)

brandon-ja commented on utils.py:
```diff
@@ -0,0 +1,56 @@
+    if len(search_term) > SEARCH_TERM_MAX_LEN:
+        raise Exception(f"search_term length is {len(search_term)}. It can not be longer than {SEARCH_TERM_MAX_LEN}")
+
+    wordlist = search_term.lower().split()
```
are you sure that the operators are case insensitive? is `AND` the same as `and` or are they treated differently by the github search api?

heger2 replied:
I checked and I couldn't find it in the doc, but from testing they aren't case sensitive. If you want to search for the work "and" then just surround the whole search term in quotes seems to do the trick e.g. "this and that" would search for that term not this AND that

brandon-ja commented on utils.py:
```diff
@@ -0,0 +1,56 @@
+        raise Exception(f"search_term contains too many operators. Max supported is {SEARCH_TERM_MAX_OPERATORS}")
+
+
+def search_code(search_term, user=None, per_page=30, page=1, uri="https://api.github.com/search/code"):
```
why provide a default user of `None`? if the api requires a user then the caller needs to provide one or you should provide a more meaningful default. Also, why include the URI as a parameter on this function? Why not use a file constant instead?

heger2 replied:
As discussed the file constant isn't used so the needed vars that the function uses is more self documented. I've also updated the function def to take in a "qualifiers" arg which is a dictionary of qualifiers so overall it's more flexible. The can now search on user, org, and repository

brandon-ja commented on utils.py:
```diff
@@ -0,0 +1,56 @@
+    _check_search_validity(search_term)
+
+    if per_page > 100:
+        raise Exception(f"per_page can not be more than 100 but {per_page} was passed in")
```
why not check this in the `_check_search_validity` function since that is validating the search term arg already.

brandon-ja commented on utils.py:
```diff
@@ -0,0 +1,56 @@
+    # NOTE: it looks like the api requires a user
+    # org or repository, so we should probably require
+    # that a user is passed in or at least some other
+    # qualifier https://docs.github.com/en/github/searching-for-information-on-github/searching-on-github/searching-code#considerations-for-code-search
```
you should probably raise an exception here if the user is missing and return the information you provide in this comment. It would make the api more user friendly/ faster.

brandon-ja commented on utils.py:
```diff
@@ -0,0 +1,56 @@
+
+    results = response.json()
+
+    items = [schema.Result(file_url=x.get("html_url"), repo=x.get("repository", {}).get("html_url")) for x in response.json().get("items", [])]
```
This is a super fancy list comprehension but it's not very scalable/ maintainable. If we need to add more data, change the data structure or github data changes it could become difficult to work with.

heger2 replied:
This has been broken out into smaller pieces to make it more readable

brandon-ja commented on utils.py:
```diff
@@ -0,0 +1,56 @@
+    # check if pagination has happened.
+    # NOTE: left out to avoid rate limits
+    # Could pass in number of requests and check that against rate limit
+    #if len(items) >= per_page:
```
so this is super nit picky and not important but... spacing #` `if

heger2 replied:
fixed :)

brandon-ja commented on client.py:
```diff
@@ -0,0 +1,28 @@
+import githubsearch_pb2 as schema
+import githubsearch_pb2_grpc
+
+SEARCH_SERVICE_URI = os.environ.get("GITHUB_SEARCH_SERVICE_URI", "localhost:50051")
```
why use an environment variable here rather than a yaml configuration file?

heger2 replied:
This is now able to be set via a cli arg, env var, or yaml file

brandon-ja commented on client.py:
```diff
@@ -0,0 +1,28 @@
+def run():
+    with grpc.insecure_channel(SEARCH_SERVICE_URI) as channel:
+        stub = githubsearch_pb2_grpc.GithubSearchServiceStub(channel)
+        response = stub.Search(schema.SearchRequest(search_term='prometheus', user='microsoft'))
```
this is great but maybe allowing users to pass in a search term to the client could be helpful for debugging?

heger2 replied:
This is now able to be set via a cli arg, env var, or yaml file

brandon-ja commented on server.py:
```diff
@@ -0,0 +1,32 @@
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+    githubsearch_pb2_grpc.add_GithubSearchServiceServicer_to_server(SearchService(), server)
+    server.add_insecure_port(f"[::]:{SERVER_PORT}")
+    print(f"Starting grpc server on port {SERVER_PORT}")
```
use logging here

heger2 replied:
Done

brandon-ja commented on server.py:
```diff
@@ -0,0 +1,32 @@
+    server.add_insecure_port(f"[::]:{SERVER_PORT}")
+    print(f"Starting grpc server on port {SERVER_PORT}")
+    server.start()
+    print("Server started")
```
use logging here

heger2 replied:
Done

brandon-ja commented on utils.py:
```diff
@@ -0,0 +1,56 @@
+SEARCH_TERM_MAX_LEN = 256
+SEARCH_TERM_MAX_OPERATORS = 5
+
+def _check_search_validity(search_term):
```
maybe move search_tearm_max_len and search_term_max_operators into the function signature.

heger2 replied:
Done

jhegerhorst-vivint committed:
Refactor based off of PR feedback

- utils.py moved to githubsearch.py
- docs linked in comments
- passing args to server.py and client.py can be a cli arg, env var, or from a config file
  - readme updated to reflect changes
- moved module vars to function definiion
- changed print statements to use logging lib instead
